opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The OpenTelemetry tracing API describes the classes used to generate
+distributed traces.
+
+The :class:`.Tracer` class controls access to the execution context, and
+manages span creation. Each operation in a trace is represented by a
+:class:`.Span`, which records the start, end time, and metadata associated with
+the operation.
+
+This module provides abstract (i.e. unimplemented) classes required for
+tracing, and a concrete no-op ``BlankSpan`` that allows applications to use the
+API package alone without a supporting implementation.
+
+The tracer supports implicit and explicit context propagation. By default spans
+are created as children of the currently active span, and the newly-created
+span becomes the new active span::
+
+    from opentelemetry.sdk.trace import tracer
+
+    # Create a new root span, set it as the current span in context
+    with tracer.span("parent"):
+        # Attach a new child and update the current span
+        with tracer.span("child"):
+            do_work():
+        # Close child span, set parent as current
+    # Close parent span, set default span as current
+
+Under explicit context propagation there is no concept of an active span, and
+the caller is responsible for managing the span's lifetime::
+
+    from opentelemetry.sdk.trace import tracer
+    from your.integration import deserialize_span
+
+    parent = deserialize_span(serialized_span)
+    # Explicit parent span assignment
+    with tracer.span("child", parent=parent) as child:
+        do_work(span=child)
+
+Applications should generally use a single global tracer, and use either
+implicit or explicit context propagation consistently throughout.
```

Comment:
How about libraries? The application could use explicit propagation consistently, while calling into a library which expects implicit context propagation.

Reply:
Yeah, we definitely should think about and document best practices for what libs should do. For that question, see https://github.com/open-telemetry/opentelemetry-python/pull/11/files#r290703075. Another question is what tracer they should use (e.g.: Use global tracer? Pass tracer at every call? Provide way to set library-wide tracer? Provide way to set a library-wide callback function that it calls to get the tracer?)

Reply:
Best-practices aside, users should have the final thing to say - if they, for some reason, need to 'mix' these modes (as you guys call it), that should be supported IHMO.

(As for the other questions, I'd say using a global `Tracer` would be a good thing to have)

Reply:
> The application could use explicit propagation consistently, while calling into a library which expects implicit context propagation.

This is a problem since it would mean that passing an explicit context should change the implicit one. This would be unexpected in the application, but required if it's calling into a library that uses implicit context.

> Provide way to set a library-wide callback function that it calls to get the tracer?

I thought implementations would provide a global tracer, do you want the callback so integrations can change this?